pkg/applications/tbtc/tbtc.go

```diff
@@ -0,0 +1,303 @@
+	RetrieveSignerPubkey(depositAddress string) error
+}
+
+// TBTCSystem is an interface that provides ability to interact
+// with TBTCSystem contract.
+type TBTCSystem interface {
+	// OnDepositCreated installs a callback that is invoked when an
+	// on-chain notification of a new deposit creation is seen.
+	OnDepositCreated(
+		handler func(depositAddress string),
+	) (subscription.EventSubscription, error)
+
+	// OnDepositRegisteredPubkey installs a callback that is invoked when an
+	// on-chain notification of a deposit's pubkey registration is seen.
+	OnDepositRegisteredPubkey(
+		handler func(depositAddress string),
+	) (subscription.EventSubscription, error)
+}
+
+// InitializeExtensions initializes extensions specific to the TBTC application.
+func InitializeExtensions(ctx context.Context, handle Handle) error {
+	logger.Infof("initializing tbtc extensions")
+
+	manager := &extensionsManager{handle}
+
+	err := manager.initializeRetrievePubkeyExtension(ctx)
+	if err != nil {
+		return fmt.Errorf(
+			"could not initialize retrieve pubkey extension: [%v]",
+			err,
+		)
+	}
+
+	logger.Infof("tbtc extensions have been initialized")
+
+	return nil
+}
+
+type depositEventHandler func(deposit string)
+
+type depositEventHandlerSetup func(
+	handler depositEventHandler,
+) (subscription.EventSubscription, error)
+
+type keepClosedHandlerSetup func(deposit string) (
+	keepClosedChan chan struct{},
+	keepClosedCancel func(),
+	err error,
+)
+
+type depositTransactionSubmitter func(deposit string) error
+
+type extensionsManager struct {
+	handle Handle
+}
+
+func (em *extensionsManager) initializeRetrievePubkeyExtension(
+	ctx context.Context,
+) error {
+	logger.Infof("initializing retrieve pubkey extension")
+
+	subscription, err := em.initializeExtension(
+		"retrieve pubkey",
+		func(handler depositEventHandler) (subscription.EventSubscription, error) {
+			return em.handle.OnDepositCreated(handler)
+		},
+		func(handler depositEventHandler) (subscription.EventSubscription, error) {
+			return em.handle.OnDepositRegisteredPubkey(handler)
+		},
+		em.setupKeepClosedHandler,
+		func(deposit string) error {
+			return em.handle.RetrieveSignerPubkey(deposit)
+		},
+		150*time.Minute,
+	)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		<-ctx.Done()
+		subscription.Unsubscribe()
+		logger.Infof("retrieve pubkey extension has been disabled")
+	}()
+
+	logger.Infof("retrieve pubkey extension has been initialized")
+
+	return nil
+}
+
+// TODO:
+//  1. Filter incoming events by operator interest.
+//  2. Incoming events deduplication.
+//  3. Resume extensions executions after client restart.
+func (em *extensionsManager) initializeExtension(
+	extensionName string,
+	triggerEventHandlerSetup depositEventHandlerSetup,
+	stopEventHandlerSetup depositEventHandlerSetup,
+	keepClosedHandlerSetup keepClosedHandlerSetup,
+	transactionSubmitter depositTransactionSubmitter,
+	timeout time.Duration,
+) (subscription.EventSubscription, error) {
```

Comment:
To be honest, I liked the previous name (`monitorAndAct`) more, especially that it's hidden inside the package and clearly expresses what we are doing in this function. I'd suggest going back to the previous name and renaming function paramters to clearly divide and express what is *monitor* and what is *act*. It was hard to me to figure out which function does what and put all the pieces together.

My proposition:
```
func (...) monitorAndAct(
  actionName string,
  monitoringStartFn watchDepositEventFn,
  monitoringStopFn watchDepositEventFn,
  keepClosed watchKeepClosedFn,
  actFn submitTxFn,
  timeout time.Duration,
)
```